contracts/AnyswapV4CallProxy.sol

```diff
@@ -110,44 +110,139 @@ abstract contract Whitelistable is MPCManageable {
     }
 }
 
-abstract contract Billable is Whitelistable {
-    event Fund(address indexed beneficiary, uint256 amount);
+interface IERC20 {
+    function balanceOf(address account) external view returns (uint256);
+    function transfer(address recipient, uint256 amount) external returns (bool);
+    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);
+}
+
+interface IBillManager {
+    function billCaller(address caller) external;
+    function billTarget(address target, uint256 cost) external;
+}
 
-    mapping(address => int256) public funds;
-    uint256 public expenses;
+contract MultichainBillManager {
+    event FundTarget(address indexed beneficiary, address indexed funder, uint256 amount);
+    event FundCaller(address indexed beneficiary, address indexed funder, address indexed token, uint256 amount);
 
-    modifier bill(address originator) {
-        uint256 gas = gasleft();
+    mapping(address => bool) public payByOriginator; // key is caller, value is if call is paid by tx originator or not
+    mapping(address => address) public callerFeeTokens; // key is caller, value is erc20 token address
+    mapping(address => uint256) public feePerCall; // fee per call
+    mapping(address => int256) public callerFunds; // caller funds (erc20 token funded)
+    mapping(address => int256) public targetFunds; // target funds (native token funded)
+
+    mapping(address => uint256) public tokenExpenses; // erc20 token used by caller
+    uint256 public expenses; // native token used by target
+
+    address public admin;
+
+    modifier onlyAdmin() {
+        require(msg.sender == admin, "must admin");
         _;
-        uint256 cost = (gas - gasleft()) * tx.gasprice;
-        funds[originator] -= int256(cost); // potential insufficient balances
+    }
+
+    constructor(address _admin) {
+        admin = _admin;
+    }
+
+    function setAdmin(address _admin) external onlyAdmin {
+        admin = _admin;
+    }
+
+    function billCaller(address caller) external {
```

Comment:
we can use some temp variables to record `callerFeeTokens[caller]`, `feePerCall[caller]` `callerFunds[caller]`, and use these temp variables from memory, this way may save some gas of accessing storage space. I'm not sure, but I remember someone tells this reason in Uniswap's contract.